### PR TITLE
Add const qualifiers to regular API

### PR DESCRIPTION
**Pull request by tfalders**
Follow-up to #118; this PR adds const qualifiers to the regular API functions to match the compatibility API/cuSOLVER.

### CODE REVIEW

**Review by qjojo — APPROVED** [review]
Good to have this on both APIs, nice work.

**Review by cgmb — COMMENTED** [review]
There is no ABI change, because these are C symbols and in C the function arguments are not part of the symbol name. However, this is technically a breaking change. It may be source-incompatible for users that are assigning these functions to function pointers, as the changing the type of the parameter changes the type of the function pointer.

Do we want to consider using something like this until ROCm 6.0?

```c
#ifdef HIPSOLVER_USE_MUTABLE_PTRS
// Use non-const pointers in API declarations (deprecated).
// [rocm6] Remove in ROCm 6.0.
#define HIPSOLVER_CONST
#else
// Use const pointers in API declarations.
#define HIPSOLVER_CONST const
#endif /* HIPSOLVER_USE_MUTABLE_PTRS */
```

That would be an escape hatch for users that for whatever reason needed the old declarations. I'm not sure it's really necessary, but it's an option I wanted to put forward for consideration.

It didn't seem necessary for the compatibility API because that was defined to have an identical signature to CUDA, but I'm not sure about the rest of the hipSOLVER API.

File: clients/include/utility.hpp [hunk 1]
```diff
@@ -111,10 +111,10 @@ class hipsolver_local_handle
         hipsolverDestroy(m_handle);
     }
 
-    hipsolver_local_handle(const hipsolver_local_handle&)            = delete;
-    hipsolver_local_handle(hipsolver_local_handle&&)                 = delete;
+    hipsolver_local_handle(const hipsolver_local_handle&) = delete;
+    hipsolver_local_handle(hipsolver_local_handle&&)      = delete;
     hipsolver_local_handle& operator=(const hipsolver_local_handle&) = delete;
-    hipsolver_local_handle& operator=(hipsolver_local_handle&&)      = delete;
+    hipsolver_local_handle& operator=(hipsolver_local_handle&&) = delete;
```
[the remark this answers is not in the document]
Comment (tfalders) [reply clients/include/utility.hpp:117]: Yes, per my comment here: https://github.com/ROCmSoftwarePlatform/hipSOLVER/pull/127#issuecomment-1219980254

**Comment by tfalders** [comment]
> Do we want to consider using something like this until ROCm 6.0?

That sounds like a good idea. I'll probably rename `HIPSOLVER_USE_MUTABLE_PTRS` to `HIPSOLVER_USE_CONST_QUALIFIERS` and enable it by default.

A potential pain I see is that this will require our `const_cast` calls to also depend on this being defined or not, and if it's applied to the regular API but not the compatibility API then I will need to add `const_cast` calls back to the compatibility API.

**Comment by jzuniga-amd** [comment]
This could get messy easily. I would suggest that we do not push this to 5.3, but let it mature and when we are happy, then add it to 5.4. 
--5.3 is not really broken without this change 
--We also have to think whether we want the changes to go into rocSOLVER so we don't need to cast away the const before calling rocsolver APIs

**Comment by jzuniga-amd** [comment]
Also, they asked last week to have the changelog for 5.3 ready for cherry-picking. The release will happen soon; not sure if we have time to follow all the processes for this to be merged into the release branch.

**Comment by tfalders** [comment]
Due to time constraints, it looks like we won't be able to get this into 5.3. I'll close it for now, and we can take another look later on.

**Pull request closed without merging**